### PR TITLE
data and jobs filters

### LINKED ISSUE #1439: Add a search bar to the data table
Users wish to search within our data and submission list tables. They should be able to enter text into a search bar and see a filtered list of rows. The filter should match text in any column of the table. FireCloud provided this functionality. Users liked the behavior, and we should replicate it. 

Isaac recently added a similar to filter bar to the workflow list in the submission details page.

### LINKED ISSUE #1443: Add search bar to job history table
This issue is to add a filter bar to the submission list (broken out from #1439).

### PR DESCRIPTION
**Pull request by zarsky-broad**
resolves #1439 
resolves #1443

### CODE REVIEW

File: src/components/input.js
Comment (katevoss) [inline src/components/input.js:79]: What does `onKeyDown` do? I tried searching the docs but they aren't clear.

File: src/pages/workspaces/workspace/Data.js
Comment (katevoss) [inline src/pages/workspaces/workspace/Data.js:309]: Why does the Ref Data table use hooks and the regular data table uses traditional state for setting the filter?

File: src/pages/workspaces/workspace/JobHistory.js
Comment (katevoss) [inline src/pages/workspaces/workspace/JobHistory.js]: Is our convention to coerce this into a boolean?
```suggestion
        !!filteredSubmissions && !!filteredSubmissions.length && h(AutoSizer, [
```

**Review by katevoss — COMMENTED** [review]
Comments:

1. Filtering in the Tables section works differently than the Reference Data. Reference Data filters as you type, and Tables filters after you press enter.

2. Nit: If you filter out all the Jobs, it says "No jobs run", even though jobs have been run. It's a bit misleading:
Screen Shot 2019-04-18 at 11 43 11 AM

I've tested this pretty thoroughly and it's looking good! My main outstanding question is the UX for filtering, whether it should filter on Enter or on keystroke.

**Comment by zarsky-broad** [comment]
@katevoss the requirement to hit enter with the data table is because there can be a huge amount of data there, so the filter is performed server-side. It could happen on typing, but that could wind up being a huge hit on the backend. If we want to do that, we need a larger conversation, so I did this for now.

File: src/components/input.js
Comment (zarsky-broad) [reply src/components/input.js:79]: The event fires every time a key is pressed, whether it changes the value of the input or not.

File: src/pages/workspaces/workspace/Data.js
Comment (zarsky-broad) [reply src/pages/workspaces/workspace/Data.js:309]: Because the components they're being added into do.

**Review by katevoss — APPROVED** [review]
Isaac addressed my questions, it looks good!

File: src/pages/workspaces/workspace/JobHistory.js
Comment (zarsky-broad) [inline src/pages/workspaces/workspace/JobHistory.js:102]: Before I merge this, I'd like to know if you think this is a terrible idea, @panentheos

File: src/components/input.js
Comment (panentheos) [inline src/components/input.js]: This seems fragile. Since the search event also includes the value, could we just call `onSearch` so the caller only sees one event at a time?

File: src/components/input.js
Comment (panentheos) [inline src/components/input.js:65]: Rather than explicitly getting the DOM node, we could just attach the ref directly and then set the event listener in an effect.

File: src/pages/workspaces/workspace/Data.js
Comment (panentheos) [inline src/pages/workspaces/workspace/Data.js]: nit: Why not just optionally call `setState` again?

File: src/pages/workspaces/workspace/Data.js
Comment (panentheos) [inline src/pages/workspaces/workspace/Data.js]: If this is the border color we want on these, perhaps we should roll it into the common component.

File: src/pages/workspaces/workspace/JobHistory.js
Comment (panentheos) [reply src/pages/workspaces/workspace/JobHistory.js:102]: Assuming there were performance issues with transforming this on the fly, this seems like a reasonable optimization. We could also `memoize`, but I think that's largely equivalent.

File: src/pages/workspaces/workspace/JobHistory.js
Comment (panentheos) [inline src/pages/workspaces/workspace/JobHistory.js]: `_.flatMap` might be natural fit here.

File: src/components/DataTable.js
Comment (panentheos) [inline src/components/DataTable.js:90]: TOL: It seems like we're actually introducing two new components in this change, 1. a text box that happens to be `type=search`, and 2. a special search input with lazy callback behavior and an explicit search affordance. All the complexity is around that second case, which is (currently) only used here. I wonder if we should make that an explicit distinction, so that the other inputs don't need to pull in all the underlying complexity.

Just a thought, not necessarily something to do right now.

File: src/components/DataTable.js
Comment (panentheos) [inline src/components/DataTable.js]: TOL: This is fine, though if this was structured such that the confirmed search value was written to state, we could roll it into the unified fetching logic in `componentDidMount`. That pattern is nice because it decouples the 'effect' from the state management.

File: src/components/DataTable.js
Comment (panentheos) [inline src/components/DataTable.js]: In the 'Select Data' modal when configuring a workflow, this is lacking a margin.

**Review by panentheos — APPROVED** [review]

**Pull request merged**
